axon-cosmosdb/src/main/java/org/axonframework/extensions/cosmosdb/eventsourcing/tokenstore/CosmosTokenStore.java

```diff
@@ -0,0 +1,516 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.extensions.cosmosdb.eventsourcing.tokenstore;
+
+import com.azure.cosmos.CosmosAsyncClient;
+import com.azure.cosmos.CosmosAsyncContainer;
+import com.azure.cosmos.CosmosAsyncDatabase;
+import com.azure.cosmos.CosmosException;
+import com.azure.cosmos.models.CosmosItemRequestOptions;
+import com.azure.cosmos.models.CosmosItemResponse;
+import com.azure.cosmos.models.PartitionKey;
+import org.axonframework.common.AxonConfigurationException;
+import org.axonframework.common.BuilderUtils;
+import org.axonframework.eventhandling.Segment;
+import org.axonframework.eventhandling.TrackingToken;
+import org.axonframework.eventhandling.tokenstore.ConfigToken;
+import org.axonframework.eventhandling.tokenstore.TokenStore;
+import org.axonframework.eventhandling.tokenstore.UnableToClaimTokenException;
+import org.axonframework.eventhandling.tokenstore.UnableToInitializeTokenException;
+import org.axonframework.eventhandling.tokenstore.UnableToRetrieveIdentifierException;
+import org.axonframework.serialization.SerializedObject;
+import org.axonframework.serialization.Serializer;
+import org.axonframework.serialization.SimpleSerializedObject;
+import reactor.core.publisher.Mono;
+
+import java.lang.management.ManagementFactory;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.TemporalAmount;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.IntStream;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import static java.util.Objects.isNull;
+import static org.axonframework.common.BuilderUtils.assertNonEmpty;
+import static org.axonframework.common.BuilderUtils.assertNonNull;
+
+/**
+ * An implementation of the {@link TokenStore} that allows you to store and retrieve tracking tokens with Cosmos DB.
+ *
+ * @author Gerard Klijs-Nefkens
+ * @since 4.9.0
+ */
+public class CosmosTokenStore implements TokenStore {
+
+    private static final String ID_PARTITION_KEY = "/id";
+    private static final String STORAGE_IDENTIFIER_NAME = "id";
+    private static final String STORAGE_IDENTIFIER_ID = "42";
+    private static final String ALL_QUERY = "SELECT * FROM root i";
+    private static final String NO_CURRENT_ITEM_MESSAGE = "Could not claim token.\n" +
```

Comment:
Shouldn't this message state it couldn't find the token? Furthermore, shouldn't that be followed up by the fact it couldn't find it because it hasn't been initialized yet?

Reply:
Yes, or the database name is wrong. Added those suggestions.

Reply:
The first two sentences feel a little incoherent right now. As if we first think claiming doesn't work, but second that we couldn't find it. Perhaps it is slightly clearer if we merge both in a single sentence?